src/esp/geo/VoxelGrid.cpp

```diff
@@ -0,0 +1,340 @@
+  }
+}
+
+void VoxelGrid::addVoxelToMeshPrimitives(std::vector<Mn::Vector3>& positions,
+                                         std::vector<Mn::Vector3>& normals,
+                                         std::vector<Mn::Color3>& colors,
+                                         std::vector<Mn::UnsignedInt>& indices,
+                                         const Mn::Vector3i& local_coords,
+                                         const std::vector<bool>& neighbors,
+                                         const Mn::Color3& color) {
+  // Using the data of a cubeSolid to create the voxel cube
+  assert(neighbors.size() >= 6);
+  Mn::Trade::MeshData cubeData = Mn::Primitives::cubeSolid();
+
+  // add cube to mesh
+  // midpoint of a voxel
+  Mn::Vector3 mid = getGlobalCoords(local_coords);
+
+  unsigned int sz = positions.size();
+  auto cubePositions = cubeData.positions3DAsArray();
+  auto cubeNormals = cubeData.normalsAsArray();
+  auto cubeIndices = cubeData.indices();
+
+  // TODO: Only add neccessary faces (based on neighborhood)
+  for (int i = 0; i < 24; i++) {
+    Mn::Vector3 vertOffset = cubePositions[i] * m_voxelSize / 2;
+    positions.push_back(vertOffset + mid);
+    // Set the normals to be weighted such that cubes look slightly curved
+    normals.push_back(cubePositions[i].normalized() * 1 / 4 +
+                      cubeNormals[i].normalized() * 3 / 4);
+    colors.push_back(color);
+  }
+
+  for (int i = 0; i < 6; i++) {
+    if (!neighbors[i]) {
+      for (int j = 0; j < 6; j++) {
+        indices.push_back(sz + cubeIndices[i * 6 + j][0]);
+      }
+    }
+  }
+}
+
+void VoxelGrid::addVectorToMeshPrimitives(std::vector<Mn::Vector3>& positions,
+                                          std::vector<Mn::Vector3>& normals,
+                                          std::vector<Mn::Color3>& colors,
+                                          std::vector<Mn::UnsignedInt>& indices,
+                                          const Mn::Vector3i& local_coords,
+                                          const Mn::Vector3& vec) {
+  Mn::Trade::MeshData cubeData = Mn::Primitives::cubeSolid();
+
+  Mn::Vector3 mid = getGlobalCoords(local_coords);
+  Mn::Vector3 pos1 = vec.normalized() * m_voxelSize * 1 / 3 + mid;
+  Mn::Vector3 orthog1 = Mn::Math::cross(vec, Mn::Vector3(0, 1, 0));
+  if (orthog1 == Mn::Vector3(0, 0, 0)) {
+    orthog1 = Mn::Vector3(1, 0, 0);
+  }
+  Mn::Vector3 orthog2 = Mn::Math::cross(vec, orthog1);
+
+  Mn::Vector3 pos2 = mid + orthog1.normalized() * m_voxelSize * 1 / 15;
+  Mn::Vector3 pos3 = mid + orthog2.normalized() * m_voxelSize * 1 / 15;
+  Mn::Vector3 pos4 = mid - orthog1.normalized() * m_voxelSize * 1 / 15;
+  Mn::Vector3 pos5 = mid - orthog2.normalized() * m_voxelSize * 1 / 15;
+
+  unsigned int sz = positions.size();
+  positions.push_back(pos1);
+  positions.push_back(pos2);
+  positions.push_back(pos3);
+  positions.push_back(pos4);
+  positions.push_back(pos5);
```

Comment:
General comment for all the code in this function: it's possible to do 

```cpp
positions.insert(positions.end(), {
    pos1,
    pos2,
    ...
})
``` 

and it's also much faster that way.